src/DynamicDiscreteSamplers.jl

```diff
@@ -194,7 +194,7 @@ Base.setindex!(w::Weights, v, i::Int) = (_setindex!(w.m, Float64(v), i); w)
             x = rand(rng, UInt64)
             # p_stage = significand_sum << shift & ...00000.111111...64...11110000
             shift += 64
-            target = significand_sum << shift % UInt64
+            target = _convert(UInt64, significand_sum << shift)
```

Comment:
The `%` here is necessary, there are cases when this overflows and the correct behavior is to use only the low 64 bits. This can happen when the significand sum is wider than 64 bits. Unfortunately, this is also a case that is very hard to cover with tests because of the guarantee that it will never run frequently. Here's a test that catches this, though:

```julia
w = DynamicDiscreteSamplers.FixedSizeWeights(2046*2048)
w .= repeat(ldexp.(1.0, -1022:1023), inner=2048)
w[(2046-16)*2048+1:2046*2048] .= 0
@test w.m[4] < 2.0^32*1.1 # Confirm that we created an interesting condition
f(w,n) = sum(rand(w) for _ in 1:n)
@test f(w, 2^28) ≈ 4.1543685e6*2^28 rtol=1e-6
```

This test only fails on this PR and when DEBUG is true.

Reply:
good catch, thank you, indeed I was a bit unsure on this